ulib/libax/src/cbindings/socket.rs

```diff
@@ -0,0 +1,406 @@
+        }
+    }
+
+    pub fn recv(&self, buf: &mut [u8]) -> LinuxResult<usize> {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.recv_from(buf).map(|e| e.0)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.recv(buf)?),
+        }
+    }
+
+    fn bind(&mut self, addr: SocketAddr) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.bind(addr)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.bind(addr)?),
+        }
+    }
+
+    fn connect(&mut self, addr: SocketAddr) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => Ok(udpsocket.connect(addr)?),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.connect(addr)?),
+        }
+    }
+
+    fn sendto(&self, buf: &[u8], addr: SocketAddr) -> LinuxResult<usize> {
+        match self {
+            // diff: must bind before sendto
+            Socket::Udp(udpsocket) => Ok(udpsocket.send_to(buf, addr)?),
+            Socket::Tcp(_) => Err(LinuxError::EISCONN),
+        }
+    }
+
+    fn recvfrom(&self, buf: &mut [u8]) -> LinuxResult<(usize, Option<SocketAddr>)> {
+        match self {
+            // diff: must bind before recvfrom
+            Socket::Udp(udpsocket) => {
+                Ok(udpsocket.recv_from(buf).map(|res| (res.0, Some(res.1)))?)
+            }
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.recv(buf).map(|res| (res, None))?),
+        }
+    }
+
+    fn listen(&mut self) -> LinuxResult {
+        match self {
+            Socket::Udp(_) => Err(LinuxError::EOPNOTSUPP),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.listen()?),
+        }
+    }
+
+    fn accept(&mut self) -> LinuxResult<TcpSocket> {
+        match self {
+            Socket::Udp(_) => Err(LinuxError::EOPNOTSUPP),
+            Socket::Tcp(tcpsocket) => Ok(tcpsocket.accept()?),
+        }
+    }
+
+    fn shutdown(&self) -> LinuxResult {
+        match self {
+            Socket::Udp(udpsocket) => {
+                udpsocket.peer_addr()?;
+                udpsocket.shutdown().unwrap();
+                Ok(())
+            }
+
+            Socket::Tcp(tcpsocket) => {
+                tcpsocket.peer_addr()?;
+                tcpsocket.shutdown().unwrap();
+                Ok(())
+            }
+        }
+    }
+}
+
+fn as_c_sockaddr(addr: &SocketAddr) -> ctypes::sockaddr {
+    return unsafe {
+        *(&ctypes::sockaddr_in {
+            sin_family: ctypes::AF_INET as u16,
+            sin_port: addr.port.to_be(),
+            sin_addr: ctypes::in_addr {
+                s_addr: u32::from_be_bytes(addr.addr.as_bytes().try_into().unwrap()).to_be(),
+            },
+            sin_zero: [0; 8],
+        } as *const _ as *const ctypes::sockaddr)
+    };
+}
+
+fn from_c_sockaddr(
+    addr: *const ctypes::sockaddr,
+    addrlen: ctypes::socklen_t,
+) -> LinuxResult<SocketAddr> {
+    if addrlen != size_of::<ctypes::sockaddr>() as u32 || addr.is_null() {
+        return Err(LinuxError::EINVAL);
```

Comment:
All null pointers should return `EFAULT`, and all wrong `addrlen` return `EINVAL`.